File: src/com/leozhi/topic116/Solution03.java
package com.leozhi.topic116;

/**
 * @author leozhi
 * 递归 + 使用已建立的 next 指针
 * 通过
 * 0ms
 */
public class Solution03 {
    public Node connect(Node root) {
        if (root == null) {
            return null;
        }
        if (root.left != null) {
            root.left.next = root.right;
            root.right.next = root.next != null ? root.next.left : null;
            connect(root.left);
            connect(root.right);
        }
        return root;
    }
}

File: src/com/leozhi/topic102/Solution03.cpp
#include <vector>
#include <queue>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

/**
 * @author leozhi
 * 广度优先搜索
 * 通过
 * 8ms
 */
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> res;
        if (!root) {
            return res;
        }
        queue<TreeNode*> que;
        que.push(root);
        while (!que.empty()) {
            int len = que.size();
            res.emplace_back();
            for (int i = 0; i < len; ++i) {
                auto node = que.front();
                que.pop();
                res.back().push_back(node->val);
                if (node->left) {
                    que.push(node->left);
                }
                if (node->right) {
                    que.push(node->right);
                }
            }
        }
        return res;
    }
};